### PR TITLE
Validators

### PR DESCRIPTION
**Pull request by evanmcc**
This is the main PR for the validators feature described in HIP 25: https://github.com/helium/HIP/blob/master/0025-validators.md

It is not entirely feature complete, but is close enough to start review.

Depends on: https://github.com/helium/blockchain-core/pull/727 and https://github.com/helium/proto/pull/59

### CODE REVIEW

**Comment by nbahramzad** [comment]
As someone that would like to run a Validator on an arm64 server, could you please include the appropriate Dockerfile and build pipeline files?

File: src/miner_val_heartbeat.erl
Comment (vihu) [inline src/miner_val_heartbeat.erl:81]: Just wondering if the fire and forget here for the heartbeat txn is sufficient enough, do we need to do `submit/2`, supply a callback and resend if the heartbeat txn fails say X number of times?

File: test/miner_blockchain_SUITE.erl
Comment (vihu) [inline test/miner_blockchain_SUITE.erl:25]: I'm only nitpicking but can we have an entirely separate suite with the three tests here if it's not too much pain? `miner_validator_SUITE` perhaps?

File: src/miner_val_heartbeat.erl
Comment (Vagabond) [reply src/miner_val_heartbeat.erl:81]: This loop might spam a lot of redundant transactions as-is, it might be better to use the callback to wait for the one submitted to clear/fail instead of trying every single block.

**Comment by andymck** [comment]
lgtm.  On a related note, i will use the added 'mode' env var to determine whether to start sibyl in miner

File: src/cli/miner_cli_genesis.erl
Comment (Vagabond) [inline src/cli/miner_cli_genesis.erl:195]: Remove?

File: src/cli/miner_cli_genesis.erl
Comment (Vagabond) [inline src/cli/miner_cli_genesis.erl:396]: This should probably be 3 once we merge the intermediate unstake version?

File: src/miner.app.src
Comment (Vagabond) [inline src/miner.app.src:10]: Why?

File: src/miner.erl
Comment (Vagabond) [inline src/miner.erl:736]: This intentional?

File: src/miner_util.erl
Comment (Vagabond) [inline src/miner_util.erl:91]: Since we don't use the gps information at all, can we just always send the 'release version'?

File: src/miner_consensus_mgr.erl
Comment (Vagabond) [inline src/miner_consensus_mgr.erl:136]: Remind me why this changed?

Comment (Vagabond) [reply src/miner_consensus_mgr.erl:136]: Are these just no longer possible?

File: src/miner_val_heartbeat.erl
Comment (Vagabond) [inline src/miner_val_heartbeat.erl:61]: What is the purpose of the waiting state? just to avoid doing extra checks for a few blocks?

File: src/miner_val_heartbeat.erl
Comment (Vagabond) [inline src/miner_val_heartbeat.erl:105]: Would it not be better to look at the result here and also use the `validator_liveness_interval` to inform the txn_wait?

File: test/miner_blockchain_SUITE.erl
Comment (Vagabond) [reply test/miner_blockchain_SUITE.erl:26]: Yes, we should have these more broken up, but that's probably out of scope for this right now.

File: test/miner_blockchain_SUITE.erl
Comment (Vagabond) [inline test/miner_blockchain_SUITE.erl:315]: Are these the transactions clearing??

File: rebar.config
Comment (vihu) [inline rebar.config:13]: reminder, switch to master when branch lands. also looks like no corresponding PR on core side?

File: src/handlers/miner_dkg_penalty_handler.erl
Comment (vihu) [inline src/handlers/miner_dkg_penalty_handler.erl:97]: nit: indentation

File: src/handlers/miner_dkg_penalty_handler.erl
Comment (vihu) [inline src/handlers/miner_dkg_penalty_handler.erl:102]: I'm wondering whether this node should multicast `{conf, GoodSigs}` even if it sent conf before when it sees `SigLen >= Majority`, mostly because it would help propagate conf msgs faster, I _think_. If I'm entirely mistaken, can we add a little comment here with some reasoning plz?

File: src/handlers/miner_dkg_penalty_handler.erl
Comment (vihu) [inline src/handlers/miner_dkg_penalty_handler.erl:84]: Maybe it doesn't matter but the order of operation here is flipped I think, membership check in state.members should happen first imo

File: src/handlers/miner_dkg_penalty_handler.erl
Comment (vihu) [inline src/handlers/miner_dkg_penalty_handler.erl:28]: Uhm, is this module missing handling `tc_key_share:tc_key_share()` entirely?

File: test/miner_blockchain_SUITE.erl
Comment (evanmcc) [reply test/miner_blockchain_SUITE.erl:26]: since these are redundant now (they're now tested in core), I'm going to delete all but the transition test.

File: src/miner.erl
Comment (evanmcc) [reply src/miner.erl:736]: yes

File: src/miner.app.src
Comment (evanmcc) [reply src/miner.app.src:10]: removing

File: src/miner_util.erl
Comment (evanmcc) [reply src/miner_util.erl:91]: this is a good idea, but we'd need to change it to read lsb-release on the gateways

File: src/miner_consensus_mgr.erl
Comment (evanmcc) [reply src/miner_consensus_mgr.erl:136]: yeah, that was an ancient clause for backwards compatibility that I finally got rid of

File: src/miner_val_heartbeat.erl
Comment (evanmcc) [reply src/miner_val_heartbeat.erl:61]: so we don't spam a new txn while waiting for the old one to land

File: src/miner_val_heartbeat.erl
Comment (evanmcc) [reply src/miner_val_heartbeat.erl:105]: I don't actually need to look at the response here, since if we succeeded, we won't submit another one, and if we failed, we'll just submit another one in the next block we get

File: rebar.config
Comment (evanmcc) [reply rebar.config:13]: most of the core stuff has already landed, and I think that we have landed all the subsequent relevant PRs as well.  We do need to upgrade, though.

**Review by Vagabond — APPROVED** [review]

**Pull request merged**
